onnxruntime/core/providers/cpu/tensor/upsample_aa.cc

```diff
@@ -0,0 +1,153 @@
+      int64_t xmax_cut = std::min<int64_t>(xmax_real, input_size);
+
+      xmin = exclude_outside ? xmin_cut : xmin_real;
+      xmax = exclude_outside ? xmax_cut : xmax_real;
+      bound_idx.push_back(xmin_cut);
+      bound_idx.push_back(xmax_cut);
+
+      auto* scale_buffer = &scale_data[i * window_size];
+      int32_t* scale_buffer_int = reinterpret_cast<int32_t*>(scale_buffer);
+      int64_t x = 0;
+      xmax -= xmin;
+      for (; x < xmax; x++) {
+        float w = p.filter((x + xmin - center + 0.5f) * inv_scale);
+        scale_buffer[x] = w;
+        total_weight += w;
+      }
+
+      if (!exclude_outside) {
+        int64_t neg_xsize = xmin < 0 ? -xmin : 0;
+        for (x = 0; x < neg_xsize; x++) {
+          scale_buffer[neg_xsize] += scale_buffer[x];
+        }
+        int64_t bound_xsize =
+            xmax + xmin > input_size ? xmax + xmin - input_size : 0;
+        for (x = xmax - bound_xsize; x < xmax; x++) {
+          scale_buffer[xmax - bound_xsize - 1] +=
+              scale_buffer[x];
+        }
+        for (x = 0; (neg_xsize | bound_xsize) > 0 && x < xmax_cut - xmin_cut; x++) {
+          scale_buffer[x] = scale_buffer[x + neg_xsize];
+        }
+      }
+
+      for (x = 0; x < xmax; x++) {
+        if (total_weight != 0.0 && total_weight != 1) {
+          scale_buffer[x] /= total_weight;
+        }
+
+        // normalize the scale to 1 << 22 for int8/uint8
+        if (dtype == ONNX_NAMESPACE::TensorProto_DataType_INT8 || dtype == ONNX_NAMESPACE::TensorProto_DataType_UINT8) {
+          if (scale_buffer[x] < 0) {
+            scale_buffer_int[x] =
+                (int)(-0.5 + scale_buffer[x] * (1 << 22));
+          } else {
+            scale_buffer_int[x] =
+                (int)(0.5 + scale_buffer[x] * (1 << 22));
+          }
+        }
+      }
+
+      for (; x < window_size; x++) {
+        scale_buffer[x] = 0;
+      }
+    }
+    return window_size;
+  };
+
+  const size_t width_rindex = is_nchw ? 0 : 1;
+  const size_t height_rindex = is_nchw ? 1 : 2;
+  const size_t channel_rindex = is_nchw ? 2 : 3;
+
+  /* Handles values form -640 to 639. */
+  static uint8_t clip8_lookups_table[1280];
+  p.clip8_lookups_table = clip8_lookups_table;
+
+  p.init_clip_lookup();
+  p.dim_y.window_size = compute_weight_coefficients(p, input_h_w_c[0], output_h_w_c[0], height_rindex,
+                                                    p.dim_y.bound, p.dim_y.original, scale_h_w_c[0], p.dim_y.weight_coefficients);
+  p.dim_x.window_size = compute_weight_coefficients(p, input_h_w_c[1], output_h_w_c[1], width_rindex,
+                                                    p.dim_x.bound, p.dim_x.original, scale_h_w_c[1], p.dim_x.weight_coefficients);
+  if (input_h_w_c.size() == 3) {
+    p.dim_z.window_size = compute_weight_coefficients(p, input_h_w_c[2], output_h_w_c[2], channel_rindex,
+                                                      p.dim_z.bound, p.dim_z.original, scale_h_w_c[2], p.dim_z.weight_coefficients);
+  }
```

Comment:
Can you please add a test with antialiasing and 3D input as it currently crashes due to passing in channel_rindex here. That leads to roi_start being -1 in SetupUpsampleFilterAA and an invalid access assertion in get_original_coordinate.

The call stack for 3D input is Upsample::BaseCompute -> UpsampleTrilinearAA -> SetupUpsampleFilterAA.